### PR TITLE
refactor: set default toolbar position to bottom center of the viewport

### LINKED ISSUE #22783: [Idea] Always default the toolbar to be positioned in the same place
Despite the toolbar being a very useful tool, it always seems to appear at random – in different places, and sometimes without being specifically requested. (And may inadvertently [affect people psychologically](https://posthog.slack.com/archives/C02E3BKC78F/p1717718049689369).)

Suggestion is to affix it to the bottom center of the viewport by default, every time it's loaded. People can move it around, but bottom/center seems like the least intrusive place for it to appear, and the predictability will be great.

### PR DESCRIPTION
**Pull request by namit-chandwani**
## Linked Issues
- Closes https://github.com/PostHog/posthog/issues/22783

## Problem

- The toolbar appears at random locations and times, causing unpredictability and potential inconvenience for users.
- Inconsistent appearance may negatively affect user experience and psychological comfort.
- Users may find it difficult to locate or use the toolbar effectively due to its random positioning.
- The current behavior lacks a default, predictable position, which could be improved by affixing the toolbar to the bottom center of the viewport.

## Changes

- **Backend:**
    - No changes required

- **Frontend (`ToolbarLogic.ts` & `ToolbarUtils.ts`):**
	- **Type Annotations (`ToolbarLogic.ts`):**
		- Added type annotations for `ToolbarLogicType` to improve type safety and clarity.
		
	- **Initial Position Calculation (`ToolbarUtils.ts`):**
		- Introduced `calculateInitialDragPosition` to calculate the initial drag position of the toolbar based on its dimensions and the window size.
		- Ensures the toolbar starts at the bottom center of the window.

	- **Error Handling (`ToolbarUtils.ts`):**
		- Added error handling to return `null` if the toolbar element is not found or if its dimensions are invalid.
		- Logs a console error if the toolbar element cannot be found.

	- **State Management (`ToolbarLogic.ts`):**
		- Updated state management to set `initialPosition` based on the calculated initial drag position.
		- Ensures the initial position is only set if the current position is `null`, avoiding unnecessary updates.

## Does this work well for both Cloud and self-hosted?

- Yes, will work well for both Cloud and self-hosted Posthog instances.

## How did you test this code?

-   **Backend Testing:**
    
    -   No changes involved

-   **Frontend Testing:**
    
    -   **Manual Testing:**    
	    -   Verified the toolbar consistently appears at the bottom center of the viewport upon initial load.
	    -   Resized the browser window to ensure the toolbar maintains its position correctly.
	    -   Dragged the toolbar to different positions and refreshed the page to confirm it initializes at the correct default position.
	    -   Checked the browser console for any error messages related to invalid toolbar dimensions or missing elements.

	-   **Cross-Browser Testing:**
	    -   Tested the changes in multiple browsers (e.g., Chrome, Firefox, Safari) to ensure consistent behavior across different environments.

-   **Screen Recording of Testing:**

      

https://github.com/user-attachments/assets/5542d8ad-991e-40fb-811e-74a61d568f0e

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactoring (altering code without changing its external behaviour)
- [ ] Documentation change
- [ ] Other

## Checklist:

- [x] Development completed
- [x] Comments added to code (wherever necessary)
- [ ] Documentation updated (if applicable)
- [ ] Added tests [Unit tests/Integration tests] (if required)
- [x] Tested changes locally

## Follow-up tasks (if any):

- None

## Additional Comments
-   Please let me know if any further changes are required elsewhere

### CODE REVIEW

**Review by benjackwhite — CHANGES_REQUESTED** [review]

File: frontend/src/toolbar/bar/Toolbar.tsx [hunk 1]
```diff
@@ -143,16 +143,53 @@ export function ToolbarInfoMenu(): JSX.Element | null {
...
+            return
+        }
+
+        // Calculate the initial position: bottom center of the window
+        const initialX = (window.innerWidth - toolbarWidth) / 2
+        const initialY = window.innerHeight - toolbarHeight
+
+        // Set the initial drag position
+        setDragPosition(initialX, initialY)
+    }
```
[the remark this answers is not in the document]
Comment (namit-chandwani) [reply frontend/src/toolbar/bar/Toolbar.tsx]: @benjackwhite This is been resolved via commit: https://github.com/PostHog/posthog/pull/24175/commits/055feb037fcba8c6eb8e5595f1dd9742c5fcec7c now

Request you to have a look. Thanks!

**Review by benjackwhite — APPROVED** [review]

**Comment by namit-chandwani** [comment]
Hey @benjackwhite, I've just pushed the latest commit [5519a07](https://github.com/PostHog/posthog/pull/24175/commits/5519a07e116a5a76f94a84575f60a36ede13f517), which updates the snapshots for the Toolbar story. 

This should resolve the failing Storybook visual regression test. I’ve also verified it locally by running the test using the command:

```bash
pnpm test:visual:ci:verify frontend/src/toolbar/Toolbar.stories.tsx
```

Looks like we’re good to go now!

**Comment by namit-chandwani** [comment]
Hey @benjackwhite, thank you for your support with the review and approval. 

Could you please trigger the CI to re-run the visual regression test? I’ve updated the snapshots for the Toolbar story and confirmed that the test passes locally.

As this is one of my initial PRs to PostHog, I’m excited to see it through to completion :)
I’m here to help with any remaining steps if needed.

TIA

**Comment by namit-chandwani** [comment]
@benjackwhite Looks like the storybook visual regression test is still failing.

However, after updating the snapshots ([here's the commit](https://github.com/PostHog/posthog/pull/24175/commits/5519a07e116a5a76f94a84575f60a36ede13f517)), I had ran the tests locally using:

```bash
pnpm test:visual:ci:verify frontend/src/toolbar/Toolbar.stories.tsx
```

Locally, all tests were passing without any problems, as shown in the screenshot:

Screenshot 2024-08-28 at 6 28 46 PM




Do you have any suggestions on what I might be missing or what else I can try?

Appreciate your help!

**Comment by benjackwhite** [comment]
I really can't figure out why the tests keep failing...

**Comment by benjackwhite** [comment]
This PR was stuck on my brain today and I thought there might actually be a better option... The issue with this one is that as soon as the element moves, the fixed position is lost forever. I put something together that allows "snapping" the position to different areas which I think [works a little nicer](https://github.com/PostHog/posthog/pull/24648).

Sorry for jumping in 🙈 without this PR I wouldn't have thought of the other solution for sure!

**Pull request closed without merging**